### PR TITLE
Mamba2 SSD Combined Forward Pass (Blackwell CuTe DSL Kernel)

### PR DESCRIPTION
**Pull request by ishovkun**
## 📌 Description

This PR adds a high-performance **Mamba2 Structured State-Space Duality (SSD) combined forward kernel** targeting Blackwell (SM100+) GPUs, implemented using NVIDIA's CuTe DSL. The kernel fuses the entire SSD chunk-scan forward pass — including cumulative sum preprocessing, chunk state computation, inter-chunk state passing, and output projection — into a single persistent kernel launch.

### Features

- **Datatypes**: bf16 I/O, bf16/fp16 state cache (fp32 state not yet supported)
- **D tensor fusion**: optional additive bias with per-head (`d_has_hdim=True`) or broadcast (`d_has_hdim=False`) modes
- **Z gating**: optional sigmoid gating on the output
- **Initial states**: optional user-provided initial hidden states
- **Variable-length sequences**: packed varlen via `seq_idx` + precomputed chunk metadata
- **Batched mode**: uniform-length sequences without varlen overhead
- **CUDA graph compatible**: all allocations cached, no host-device sync in the hot path

### Architecture

The kernel uses **16 warps** with warp-level task specialization:

| Warps | Role |
|-------|------|
| 0 | TMA loads (B, C, x, dt, A, D, z) |
| 1–4 | MMA: CB = C^T B (state-space matrix) |
| 5–8 | MMA: P = CB * decay (intra-chunk transitions) |
| 9 | Preprocessing (dt softplus/limit, dA cumsum, decay masks) |
| 10–13 | MMA: states = B^T (x * dt) + decay * prev_state |
| 14–15 | Epilogue MMA: out = C * state, write output via TMA |

A persistent tile scheduler distributes chunks across SMs. Shape-polymorphic compilation (via CuTe DSL fake tensors with symbolic dims) means **one compilation covers all batch/seqlen combinations**.

### Files Added/Modified

| Area | Key files |
|------|-----------|
| Python API | `flashinfer/mamba/ssd_combined.py` — `SSDCombined` class |
| CuTe DSL kernel | `flashinfer/mamba/ssd_kernel.py` (4.3k lines) |
| Tile scheduler | `flashinfer/mamba/ssd_tile_scheduler.py` |
| Cumsum preprocessing | `flashinfer/triton/kernels/ssd_chunk_state.py` (Triton) |
| Seq-chunk metadata | `include/flashinfer/mamba/seq_chunk_cumsum.cuh`, `csrc/seq_chunk_cumsum.cu` |
| JIT integration | `flashinfer/jit/mamba/seq_chunk_cumsum.py` |
| Tests | `tests/mamba/test_chunk_scan_combined.py` (1.9k lines) |
| Triton reference | `tests/mamba/triton_reference/ssd_*.py` |
| Benchmark | `benchmarks/bench_mamba_ssd_combined.py` |

### Performance

All numbers on **NVIDIA B200**, bf16, `chunk_size=128, nheads=8, headdim=64, dstate=128, ngroups=8`. The timings were collected using cuda-graphs.

**Column definitions:**
- **batch** / **num_seqs**: number of independent sequences (batched) or packed user sequences (varlen)
- **chunks/seq**: number of `chunk_size`-token chunks per sequence (`seqlen / chunk_size`)
- **total chunks**: total chunk count across all sequences (`batch * chunks/seq`)
- **total seqlen**: total number of tokens processed (`batch * chunks/seq * chunk_size`)
- **FlashInfer (ms)**: end-to-end wall time for the fused CuTe DSL kernel (including Triton cumsum preprocessing)
- **Triton (ms)**: end-to-end wall time for the Triton reference (5 separate kernel launches)
- **Speedup**: `Triton / FlashInfer` (>1x means FlashInfer is faster)

#### Batched mode (uniform sequence lengths, no initial states)

| batch | chunks/seq | total chunks | total seqlen | FlashInfer (ms) | Triton (ms) | Speedup |
|------:|----------:|-----------:|------------:|----------------:|------------:|--------:|
| 1 | 1 | 1 | 128 | 0.012 | 0.016 | 1.28x |
| 1 | 4 | 4 | 512 | 0.032 | 0.018 | 0.57x |
| 1 | 16 | 16 | 2,048 | 0.112 | 0.031 | 0.28x |
| 1 | 64 | 64 | 8,192 | 0.425 | 0.080 | 0.19x |
| 1 | 256 | 256 | 32,768 | 1.675 | 0.317 | 0.19x |
| 4 | 1 | 1 | 128 | 0.013 | 0.018 | 1.44x |
| 4 | 4 | 4 | 512 | 0.034 | 0.029 | 0.87x |
| 4 | 16 | 16 | 2,048 | 0.112 | 0.073 | 0.65x |
| 4 | 64 | 64 | 8,192 | 0.426 | 0.255 | 0.60x |
| 16 | 1 | 1 | 128 | 0.015 | 0.033 | 2.22x |
| 16 | 4 | 4 | 512 | 0.036 | 0.074 | 2.05x |
| 16 | 16 | 16 | 2,048 | 0.116 | 0.252 | 2.17x |
| 64 | 1 | 1 | 128 | 0.038 | 0.095 | 2.50x |
| 64 | 4 | 4 | 512 | 0.118 | 0.261 | 2.21x |
| 64 | 16 | 16 | 2,048 | 0.437 | 0.952 | 2.18x |
| 128 | 1 | 1 | 128 | 0.062 | 0.173 | 2.81x |
| 128 | 4 | 4 | 512 | 0.201 | 0.502 | 2.49x |
| 128 | 16 | 16 | 2,048 | 0.759 | 1.881 | 2.48x |
| 256 | 1 | 1 | 128 | 0.113 | 0.326 | 2.87x |
| 256 | 4 | 4 | 512 | 0.391 | 0.982 | 2.51x |
| 256 | 16 | 16 | 2,048 | 1.507 | 3.736 | 2.48x |
| 512 | 2 | 2 | 256 | 0.397 | 1.028 | 2.59x |

> At low batch / many chunks per sequence (batch=1, long sequences), the Triton reference is faster because it launches 5 separate small kernels with lower per-kernel overhead. The fused CuTe kernel wins at **batch >= 16** where parallelism across sequences saturates the GPU.

#### Varlen mode (packed sequences, with initial states)

| num_seqs | chunks/seq | total chunks | total seqlen | FlashInfer (ms) | Triton (ms) | Speedup |
|---------:|----------:|-----------:|------------:|----------------:|------------:|--------:|
| 1 | 1 | 1 | 128 | 0.022 | 0.027 | 1.22x |
| 4 | 1 | 4 | 512 | 0.023 | 0.030 | 1.31x |
| 8 | 1 | 8 | 1,024 | 0.024 | 0.041 | 1.71x |
| 32 | 1 | 32 | 4,096 | 0.046 | 0.118 | 2.56x |
| 64 | 1 | 64 | 8,192 | 0.080 | 0.211 | 2.63x |
| 128 | 1 | 128 | 16,384 | 0.134 | 0.388 | 2.90x |
| 256 | 1 | 256 | 32,768 | 0.259 | 0.753 | 2.91x |
| 4 | 8 | 32 | 4,096 | 0.119 | 0.103 | 0.87x |
| 8 | 8 | 64 | 8,192 | 0.121 | 0.177 | 1.46x |
| 16 | 8 | 128 | 16,384 | 0.124 | 0.335 | 2.70x |
| 32 | 8 | 256 | 32,768 | 0.237 | 0.657 | 2.77x |
| 64 | 8 | 512 | 65,536 | 0.463 | 1.283 | 2.77x |
| 32 | 32 | 1,024 | 131,072 | 0.896 | 2.486 | 2.77x |
| 64 | 32 | 2,048 | 262,144 | 1.771 | 4.936 | 2.79x |
| 128 | 32 | 4,096 | 524,288 | 3.099 | 9.819 | 3.17x |

> In the serving-relevant regime (many short sequences packed together), FlashInfer is consistently **2.5–3.2x faster** than Triton. The single-fused-kernel design amortizes launch overhead across all packed sequences.

### Reproducing Benchmarks

The benchmark script is included at `benchmarks/bench_mamba_ssd_combined.py`:

```bash
# Batched mode (CUDA graphs)
python benchmarks/bench_mamba_ssd_combined.py --batched --cuda_graph

# Varlen mode (CUDA graphs)
python benchmarks/bench_mamba_ssd_combined.py --varlen --cuda_graph
```

### Limitations / Future Work

- **fp32 state cache** not yet supported (bf16/fp16 only)
- **Forward pass only** — backward pass not included
- **SM100+ required** — Blackwell only (CuTe DSL)
- Low-batch / long-sequence regime (batch=1, many chunks) is slower than Triton due to persistent-kernel overhead vs. 5 small kernel launches

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

## Reviewer Notes

The kernel is Blackwell-only. Please check if I handled all the imports correctly.

## Summary by CodeRabbit

* **New Features**
  * New SSDCombined implementation with varlen, initial-states, optional scaling and gating, plus GPU-backed chunked primitives and a tile scheduler for faster chunked SSM/attention workloads.
* **Benchmarks**
  * Added a CLI benchmarking tool with single/multi-config sweeps and profiling modes (NCU, profiler) for per-configuration comparisons.
* **Tests**
  * Large test suite covering batched/varlen paths, dtype combinations, gating, initial-states, and end-to-end correctness.

### CODE REVIEW

File: flashinfer/mamba/ssd_tile_scheduler.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+            obj_values = extract_mlir_values(obj)
+            values += obj_values
+            self._values_pos.append(len(obj_values))
+        return values
+
+    def __new_from_mlir_values__(self, values):
+        obj_list = []
+        for obj, n_items in zip(
+            [self.problem_shape_ntiles, self.eh, self.ngroup_ratio], self._values_pos
+        ):
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply flashinfer/mamba/ssd_tile_scheduler.py:73]: No fix needed — the code already has `strict=False` explicitly set.

File: flashinfer/mamba/ssd_tile_scheduler.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+from cutlass.cutlass_dsl import (
+    Boolean,
+    Integer,
+    Int32,
+    min,
+    extract_mlir_values,
+    new_from_mlir_values,
+    dsl_user_op,
+)
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply flashinfer/mamba/ssd_tile_scheduler.py:38]: done

File: tests/mamba/triton_reference/ssd_chunk_state.py [hunk 3]
```diff
@@ -0,0 +1,771 @@
...
+    )
+    dt_ptrs = dt_ptr + offs_k * stride_dt_csize
+    dA_cs_last = tl.load(
+        dA_cumsum_ptr + (end_idx - pid_c * chunk_size - 1) * stride_dA_cs_csize
+    ).to(tl.float32)
+    dA_cumsum_ptrs = dA_cumsum_ptr + offs_k * stride_dA_cs_csize
+
+    chunk_size_limit = end_idx - pid_c * chunk_size
+    start_idx = tl.load(cu_seqlens_ptr + pid_b)
+    start_idx_cur = tl.maximum(start_idx - pid_c * chunk_size, 0)
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply tests/mamba/triton_reference/ssd_chunk_state.py:475]: This is a Triton reference implementation used only in tests — not production code. The finding describes a real edge case (zero-length sequences would cause `(end_idx - 1) // chunk_size` to go negative), but this is test reference code that's only called with well-formed inputs from the test harness. Zero-length sequences are not part of the test scenarios.

No fix needed — this is defensive hardening of test-only reference code against inputs that never occur in practice.

File: tests/mamba/triton_reference/ssd_chunk_state.py [hunk 4]
```diff
@@ -0,0 +1,771 @@
...
+            dA_cumsum_ptrs, mask=offs_k < chunk_size_limit - k, other=0.0
+        ).to(tl.float32)
+        dt_k = tl.load(dt_ptrs, mask=offs_k < chunk_size_limit - k, other=0.0).to(
+            tl.float32
+        )
+        scale = tl.where(
+            (offs_k >= start_idx_cur - k) & (offs_k < chunk_size_limit - k),
+            tl.exp(dA_cs_last - dA_cs_k) * dt_k,
+            0.0,
+        )
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply tests/mamba/triton_reference/ssd_chunk_state.py:503]: Changing the reference to mask potential numerical issues would defeat its purpose as a ground truth. No fix needed.

File: tests/mamba/triton_reference/ssd_combined.py [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+            return
+        else:
+            return final_states
+    else:
+        varlen_states = rest[0]
+        return (
+            (varlen_states)
+            if not return_final_states
+            else (final_states, varlen_states)
+        )
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply tests/mamba/triton_reference/ssd_combined.py:287]: The `return` on line 268 (returning `None`) is intentional — when `return_final_states=False` and `return_varlen_states=False`, the caller doesn't need any return value. The finding's suggestion to return `final_states` in that branch contradicts the function's semantics. No fix needed.

File: flashinfer/mamba/ssd_combined.py [hunk 6]
```diff
@@ -0,0 +1,670 @@
...
+        # Varlen metadata: seq_chunk_cumsum is output buffer for Triton kernel, passed as input to CUTLASS kernel
+        seq_chunk_cumsum = None
+        num_seqs = 0
+
+        if (
+            seq_idx is not None
+            and chunk_indices is not None
+            and chunk_offsets is not None
+        ):
+            num_seqs = initial_states.shape[0] if initial_states is not None else 1
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply flashinfer/mamba/ssd_combined.py]: varlen case always has initial states. added a check that this never happens in the last commit.

File: include/flashinfer/mamba/seq_chunk_cumsum.cuh [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+  if (s == num_seqs - 1 || (s < num_seqs && s == num_seqs - 1)) {
+    output[num_seqs] = global_prefix + inclusive;
+  }
+  // Handle case where num_seqs isn't a multiple of TILE_SIZE:
+  // the last block's last active thread writes the final element.
+  int last_seq_in_tile = min((tile_idx + 1) * TILE_SIZE, num_seqs) - 1;
+  if (s == last_seq_in_tile) {
+    output[num_seqs] = global_prefix + inclusive;
+  }
+}
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply include/flashinfer/mamba/seq_chunk_cumsum.cuh]: Thank you! fixed in the last commit!

File: flashinfer/mamba/ssd_combined.py [hunk 8]
```diff
@@ -0,0 +1,670 @@
...
+        self._has_d = has_d
+        self._d_has_hdim = d_has_hdim
+        self._has_init_states = has_initial_states
+        self._has_varlen = has_varlen
+        self._has_z = has_z
+        self._state_torch_dtype = state_dtype
+
+        # Resolve dtypes
+        assert io_dtype == torch.bfloat16, f"io_dtype must be bfloat16, got {io_dtype}"
+        self._io_dtype = cutlass.BFloat16
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply flashinfer/mamba/ssd_combined.py:308]: I updated the PR description to add that I/O is supported only in bf16.

File: include/flashinfer/mamba/seq_chunk_cumsum.cuh [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+ */
+template <typename SeqIdxT>
+__device__ __forceinline__ int lower_bound_seq_id(const SeqIdxT* __restrict__ seq_idx,
+                                                  const int32_t* __restrict__ chunk_indices,
+                                                  const int32_t* __restrict__ chunk_offsets,
+                                                  int chunk_size, int num_logical_chunks,
+                                                  int target) {
+  int lo = 0, hi = num_logical_chunks;
+  while (lo < hi) {
+    int mid = lo + (hi - lo) / 2;
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply include/flashinfer/mamba/seq_chunk_cumsum.cuh]: Fixed in the latest commit.

File: csrc/seq_chunk_cumsum.cu [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+  cudaError_t status;
+  DISPATCH_DLPACK_IDTYPE_TO_CTYPE(seq_idx.dtype(), SeqIdxT, [&] {
+    status = SeqChunkCumsumLauncher(
+        static_cast<const SeqIdxT*>(seq_idx.data_ptr()),
+        static_cast<const int32_t*>(chunk_indices.data_ptr()),
+        static_cast<const int32_t*>(chunk_offsets.data_ptr()),
+        static_cast<int32_t*>(output.data_ptr()), static_cast<int>(chunk_size),
+        static_cast<int>(num_logical_chunks), static_cast<int>(num_seqs), stream);
+    return true;
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply csrc/seq_chunk_cumsum.cu]: There is no way any of these parameters is larger than 2 billion. They are in the order of 100-10000.

File: include/flashinfer/mamba/seq_chunk_cumsum.cuh [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+    __syncthreads();
+    global_prefix = smem[TILE_SIZE];
+  }
+
+  // Write output: exclusive prefix = inclusive - count + global_prefix
+  if (s < num_seqs) {
+    output[s] = global_prefix + inclusive - count;
+  }
+  // Last active thread writes the sentinel
+  if (s == num_seqs - 1 || (s < num_seqs && s == num_seqs - 1)) {
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply include/flashinfer/mamba/seq_chunk_cumsum.cuh]: this code is outdated

File: tests/mamba/test_seq_chunk_cumsum.py [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+class TestSeqChunkCumsum:
+    """Test the seq_chunk_cumsum CUDA kernel against the Python reference."""
+
+    CHUNK_SIZE = 128
+
+    def _call_kernel(self, seq_idx, chunk_indices, chunk_offsets, chunk_size, num_seqs):
+        """Call the CUDA kernel. Will fail until the kernel is implemented."""
+        from flashinfer.mamba.ssd_combined import _get_seq_chunk_cumsum_module
+
+        module = _get_seq_chunk_cumsum_module()
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply tests/mamba/test_seq_chunk_cumsum.py:97]: resolved

File: flashinfer/mamba/ssd_combined.py [hunk 13]
```diff
@@ -0,0 +1,675 @@
...
+        x_permuted = x_reshaped.permute(4, 2, 1, 3, 0)
+
+        # z: same layout as x
+        z_permuted = None
+        if z is not None:
+            z_reshaped = z.reshape(batch, nchunks, chunk_size, nheads, headdim)
+            assert z.dtype == io_torch_dtype, (
+                f"z dtype {z.dtype} doesn't match {io_torch_dtype}"
+            )
+            z_permuted = z_reshaped.permute(4, 2, 1, 3, 0)
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply flashinfer/mamba/ssd_combined.py:581]: z and D substantially affect the behavior of the kernel. For z, a simple non-null check suffices. It typically will be there in prod.

File: include/flashinfer/mamba/seq_chunk_cumsum.cuh [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+template <typename SeqIdxT>
+__device__ __forceinline__ int lower_bound_seq_id(const SeqIdxT* __restrict__ seq_idx,
+                                                  const int32_t* __restrict__ chunk_indices,
+                                                  const int32_t* __restrict__ chunk_offsets,
+                                                  int chunk_size, int num_logical_chunks,
+                                                  int target) {
+  int lo = 0, hi = num_logical_chunks;
+  while (lo < hi) {
+    int mid = lo + (hi - lo) / 2;
+    int seq_id = static_cast<int>(seq_idx[chunk_indices[mid] * chunk_size + chunk_offsets[mid]]);
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply include/flashinfer/mamba/seq_chunk_cumsum.cuh]: hell no I'm not validating data in a high-throughput kernel

**Comment by aleozlx** [comment]
tests look all clean. but i'll leave it for more review time. i haven't reviewed

**Comment by aleozlx** [comment]
(not targeting 0.6.6)

File: include/flashinfer/mamba/seq_chunk_cumsum.cuh [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+    // Multi-block path with decoupled lookback
+    using TileState = cub::ScanTileState<int32_t>;
+
+    int num_tiles = (num_seqs + TILE_SIZE - 1) / TILE_SIZE;
+
+    // Allocate tile state
+    std::size_t temp_bytes = 0;
+    TileState::AllocationSize(num_tiles, temp_bytes);
+    void* d_temp = nullptr;
+    cudaError_t err = cudaMallocAsync(&d_temp, temp_bytes, stream);
```
Comment (yzh119) [inline include/flashinfer/mamba/seq_chunk_cumsum.cuh]: Can we pass workspace buffer allocated from python (which is large enough for the largest possible temp_bytes that you will encounter) instead of allocating memory here? We usually avoid memory allocation during llm inference.

File: include/flashinfer/mamba/seq_chunk_cumsum.cuh [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+    int init_threads = 256;
+    int init_blocks = (num_tiles + init_threads - 1) / init_threads;
+    InitTileStateKernel<<<init_blocks, init_threads, 0, stream>>>(tile_state, num_tiles);
+
+    // Launch scan kernel
+    SeqChunkCumsumKernelMultiBlock<<<num_tiles, TILE_SIZE, 0, stream>>>(
+        seq_idx, chunk_indices, chunk_offsets, output, tile_state, chunk_size, num_logical_chunks,
+        num_seqs);
+
+    cudaFreeAsync(d_temp, stream);
```
Comment (yzh119) [inline include/flashinfer/mamba/seq_chunk_cumsum.cuh]: ditto

Comment (ishovkun) [reply include/flashinfer/mamba/seq_chunk_cumsum.cuh]: I've added options to avoid extra allocations. Additionally, the user can now manually pass `seq_chunk_cumsum` and avoid that extra computation entirely.

**Comment by aleozlx** [comment]
CI flagged some errors on cuda 12.9 B200/B300


2026-03-23T19:32:53.352873Z 01O =========== 144 failed, 432 passed, 288 skipped in 232.02s (0:03:52) ===========
2026-03-23T19:32:53.352874Z 01O ❌ FAILED: tests/attention/test_cudnn_prefill.py

**Comment by aleozlx** [comment]
still trying to confirm if related

**Comment by aleozlx** [comment]
An independent PR bot run has exonerated this PR from the CI flagged issues.

https://gitlab-master.nvidia.com/dl/flashinfer/flashinfer-ci/-/pipelines/46807950

**Comment by aleozlx** [comment]
this is clear to merge as soon as bot run pipeline is fully done

**Comment by aleozlx** [comment]
clean. merging now
